README.md
> 🚀 Demo: http://petermikitsh.github.io/learn-single-spa/ 🚀

# learn-single-spa

Experiment with Single SPA

## Getting Started

Pre-req's:

- Git version >= 2.22
- For live development, visit `chrome://flags/#allow-insecure-localhost` and Enable.

- ```sh
  nvm use
  yarn install
  ```

## Two Development modes

- **Live Development**: Navigate to deployed server, and update feature app(s) to point to local servers
- **Local Development**: Traditional setup

### ⭐️ Live Development (`yarn run start:https`)

1. Go to live site: https://petermikitsh.github.io/learn-single-spa
2. In the console: `window.ENABLE_DEV_MODE()`
3. In your terminal, select your feature app with `yarn run start:https` (note the port number).
4. Select import map panel, select app to develop. For URL, enter port number and select "Apply Override".
5. Refresh the webpage.

### Local Development (`yarn run start`)

1. Run `yarn run start`
2. In the CLI, select apps to run. Import maps for the selected apps will point to your local webpack servers. All remaining apps will use staging import maps.
3. Go to https://0.0.0.0:9000/

#### Change import maps

Select environment import maps via `?env=stg` or `?env=prod`.

#### Servers

| Hostname     | Description           | Path                   |
| ------------ | --------------------- | ---------------------- |
| 0.0.0.0:9000 | Root App              | \*\*                   |
| 0.0.0.0:9001 | App1 (React App)      | /learn-single-spa/app1 |
| 0.0.0.0:9002 | App2 (Angular 11 App) | /learn-single-spa/app2 |
| 0.0.0.0:9003 | App3 (React App)      | /learn-single-spa      |
| 0.0.0.0:9004 | Util4 (Plain JS)      | N/A                    |

Each microfrontend server only serves its bundle.

## Releases

Releases are pushed to the `gh-pages` branch. In this example project, `gh-pages` is analogous to a CDN.

```
yarn run cicd
```

## Deployments

Import maps are defined on a per-enviornment basis, e.g: `importmap.stg.json` and `importmap.prod.json`, allowing apps to be deployed asynchronously. Which import map to use is computed at runtime, with highest to lowest precendence:

- Query parameter `env`
- Host inspection

1. Update the `packages/imports-maps/importmap.{env}.json` with versions to use.
2. Run `yarn run deploy`. **All** import map files will be deployed.

## Learnings

- When launching a new feature app, update the import map first before deploying the root config. Otherwise, your root config will reference an app that cannot be resolved via import map.
- Single SPA's utility module concept selects the module version at runtime via import maps. Because selection is at runtime and not compile time, there is no way to reliably type a `System.import` call. Typings are a compile time construct. See https://github.com/single-spa/single-spa.js.org/issues/361. The best workaround might be to use ES Modules (which are compile time) instead of browser modules (which are runtime).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.3](https://github.com/petermikitsh/learn-single-spa/compare/v0.1.2...v0.1.3) (2021-02-16)

**Note:** Version bump only for package root





## [0.1.2](https://github.com/petermikitsh/learn-single-spa/compare/v0.1.1...v0.1.2) (2021-02-16)


### Bug Fixes

* app2 start ([c8f980b](https://github.com/petermikitsh/learn-single-spa/commit/c8f980ba907ab4692904fa06ce4faf2e1a457440))





## [0.1.1](https://github.com/petermikitsh/learn-single-spa/compare/v0.1.0...v0.1.1) (2021-02-14)

**Note:** Version bump only for package root





# [0.1.0](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.8...v0.1.0) (2021-02-14)


### Features

* env-specific import maps ([a4cd26d](https://github.com/petermikitsh/learn-single-spa/commit/a4cd26d4f255b3784f6e381cd399ff422ef82acf))





## [0.0.8](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.7...v0.0.8) (2021-02-14)

**Note:** Version bump only for package root





## [0.0.7](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.6...v0.0.7) (2021-02-14)

**Note:** Version bump only for package root





## [0.0.6](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.5...v0.0.6) (2021-02-14)

**Note:** Version bump only for package root





## [0.0.5](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.4...v0.0.5) (2021-02-14)

**Note:** Version bump only for package root





## [0.0.4](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.3...v0.0.4) (2021-02-14)

**Note:** Version bump only for package root





## [0.0.3](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.2...v0.0.3) (2021-02-14)

**Note:** Version bump only for package root





## [0.0.2](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.1...v0.0.2) (2021-02-14)

**Note:** Version bump only for package root





## [0.0.1](https://github.com/petermikitsh/learn-single-spa/compare/v0.0.0-canary-f520365...v0.0.1) (2021-02-14)

**Note:** Version bump only for package root


scripts/start.js
#!/usr/bin/env node

const { spawn } = require("child_process");
const prompts = require("prompts");
const pidTree = require("pidtree");

const portMaps = {
  app1: 9001,
  app2: 9002,
  app3: 9003,
  util4: 9004,
};

(async () => {
  const { featureApps } = await prompts([
    {
      type: "multiselect",
      name: "featureApps",
      message: "Choose Feature App(s) to work on:",
      instructions: false,
      choices: Object.entries(portMaps).map(([appName, portNumber]) => ({
        title: `${appName} (Port ${portNumber})`,
        value: appName,
      })),
    },
  ]);

  if (!featureApps?.length) {
    console.log("🚨 No feature app(s) selected");
    process.exit();
  }

  if (!process.env.HTTPS) {
    console.log("Live Server at http://0.0.0.0:9000/");
  }

  const startProcess = spawn(
    "lerna",
    [
      "run",
      "start",
      "--scope",
      `'*/{root-config,${featureApps.join(",")}}'`,
      "--stream",
      "--parallel",
    ],
    {
      stdio: "inherit",
      env: {
        ...process.env,
        FEATURE_APP_DATA: JSON.stringify(
          featureApps.reduce((result, currFeatureApp) => {
            result[currFeatureApp] = portMaps[currFeatureApp];
            return result;
          }, {})
        ),
      },
    }
  );

  setTimeout(async () => {
    const ids = await pidTree(startProcess.pid);
    process.on("SIGINT", async () => {
      spawn("kill", ["-9"].concat(ids));
    });
  }, 2000);
})();
